Add a mechanism for getting line numbers for lexemes.

ltratt opened the pull request:
This adjusts lrlex to gain support for finding out (line, col) numbers for lexemes. The two commits hopefully make things clear.

Sarah: before we think about merging this, can you see if the altered API works OK for you? I think it's a reasonable design, but I'd like to check that before we go ahead!

[One thing that's implicit in this commit is that I've tried to structure things in a way that, in the future, we can run a lexer and parser in parallel to each other. Is that a good idea? No idea!]

ltratt committed:
Separate lexer definitions from lexers.

This is a step on the road to something bigger.

Previously the Lexer struct spat out a Vec of Lexemes, meaning that there was no
way to ask questions of Lexemes such as "what line number did you start with"
(since Lexemes are, intentionally, small at two machines words on a 64 bit
machine). The only way to do this is to have an intermediate result struct which
contains this sort of global-ish information.

This commit does precisely that: Lexer becomes LexerDef; on then makes a Lexer
(specific to a given input) from LexerDef; and ask the Lexer for a Vec of
Lexemes.

ltratt committed:
Add support for getting the line and column number of a lexeme.

snim2 commented:
Fascinating. Travis is telling me this failed, GH is telling me it passed.

snim2 commented:
Just to be clear, if this PR is merged, will another PR in lrpar be necessary, or can I go ahead and use this in rstreediff straight away?

ltratt commented:
You're right, this will require a small tweak to lrpar once it's merged. I'd forgotten that, so well reminded!

snim2 commented:
So, if I try to work this into my code I get two errors:

```rust
error[E0599]: no method named `lex` found for type `lrlex::LexerDef<u16>` in the current scope
   --> src/lib/ast.rs:768:35
    |
768 |     let mut lexemes = match lexer.lex(&input) {
    |                                   ^^^

error: method `new` is private
   --> src/lib/ast.rs:772:18
    |
772 |     lexemes.push(Lexeme::new(u16::try_from(usize::from(grm.end_term_idx())).unwrap(),
    |                  ^^^^^^^^^^^

error: aborting due to previous error(s)
```

ltratt commented:
Don't forget that Lexer is now LexerDef, so you'll need to do something like `let mut lexemes = match lexerdef.lexer(&input).lex()`.

I didn't realise you were creating `Lexeme`s yourself. Can you give me an idea why so I can work out if I should make `new` public or not?

snim2 commented:
This is how I am lexing the input files:
```rust
    // Lex input file.
    let mut lexemes = match lexer.lex(&input) {
        Ok(tokens) => tokens,
        Err(_) => return Err(ParseError::LexicalError),
    };
    lexemes.push(Lexeme::new(u16::try_from(usize::from(grm.end_term_idx())).unwrap(), input.len(), 0));

```

ltratt commented:
Ah, so it's just the end lexeme so that `lrpar` can do its thing. Hmm. OK, that's reasonable.

snim2 commented:
Nearly there, I think. Now I am getting:
```rust
error[E0599]: no method named `lex` found for type `lrlex::lexer::Lexer<'_, u16>` in the current scope
   --> src/lib/ast.rs:768:52
    |
768 |     let mut lexemes = match lexerdef.lexer(&input).lex() {
    |                                                    ^^^
```

and also the module `lrlex::lexer` is private, I think.

ltratt commented:
Oops `lex()` should be `lexemes()`. Sorry!

snim2 commented:
Ok, so to say `let (line_no, col_no) = lexer(token_id).line_and_col().ok();` I need to pass an `lrlex::lexer::Lexer` into  my `parse_ast` method, and I think `lexer` is private. The only other issue I have seems to be from lrpar:

```rust
   --> src/lib/ast.rs:779:56
    |
779 |     let pt = match parser::parse::<u16>(&grm, &stable, &lexemes) {
    |                                                        ^^^^^^^^ expected struct `lrlex::lexer::Lexeme`, found struct `lrlex::Lexeme`
    |
    = note: expected type `&std::vec::Vec<lrlex::lexer::Lexeme<u16>>`
               found type `&std::vec::Vec<lrlex::Lexeme<u16>>`
```

ltratt commented:
OK, see if that latest commit makes it work for you.

snim2 commented:
OK, I think I am just down to the issue with `lrpar` now!

ltratt commented:
Right, let me give lrpar a quick once over.

ltratt commented:
@snim2 In theory, with the `reflect_lrlex_changes` branch in `lrpar`, you should be good. Let me know.

ltratt committed:
Use &str instead of &String.

This simplifies the API a little bit (I thought I needed to use &String for
ownership reasons, but I was wrong).

snim2 commented on src/lib/lexer.rs:
```diff
@@ -158,6 +205,8 @@ impl<TokId: Copy> Lexeme<TokId> {
     }
 }
 
+
```
Extra lines?

ltratt replied:
Ah, let me check.

ltratt replied:
Good spot!

ltratt commented:
I suggest I squash this (probably down to 3 commits), then we merge.

ltratt commented:
OK, ready for me to squash?

snim2 commented:
Yep!

ltratt commented:
OK, squashed.
